packages/server/src/adapters/bun/createBunWSHandler.ts

```diff
@@ -0,0 +1,279 @@
+    unsubscribe(): void;
+};
+
+export function createBunWSHandler<TRouter extends AnyRouter>(
+    opts: BunWSAdapterOptions<TRouter>,
+): WebSocketHandler<BunWSClientCtx> {
+    const {router, createContext} = opts;
+
+    const respond = (
+        client: ServerWebSocket<unknown>,
+        untransformedJSON: TRPCResponseMessage,
+    ) => {
+        client.send(
+            JSON.stringify(
+                transformTRPCResponse(opts.router._def._config, untransformedJSON),
+            ),
+        );
+    };
+
+    return {
+        async open(client: ServerWebSocket<BunWSClientCtx>) {
+            const {req} = client.data;
+            const clientSubscriptions = new Map<string | number, Unsubscribable>();
+
+            const ctxPromise = createContext?.({req, res: client});
+            let ctx: inferRouterContext<TRouter> | undefined = undefined;
+            await (async () => {
+                try {
+                    ctx = await ctxPromise;
+                } catch (cause) {
+                    const error = getTRPCErrorFromUnknown(cause);
+                    opts.onError?.({
+                        error,
+                        path: undefined,
+                        type: "unknown",
+                        ctx,
+                        req,
+                        input: undefined,
+                    });
+                    respond(client, {
+                        id: null,
+                        error: getErrorShape({
+                            config: router._def._config,
+                            error,
+                            type: "unknown",
+                            path: undefined,
+                            input: undefined,
+                            ctx,
+                        }),
+                    });
+
+                    // close in next tick
+                    (global.setImmediate ?? global.setTimeout)(() => {
+                        client.close();
+                    });
+                }
+            })();
+
+            const stopSubscription = (
+                subscription: Unsubscribable,
+                {id, jsonrpc}: JSONRPC2.BaseEnvelope & { id: JSONRPC2.RequestId },
+            ) => {
+                subscription.unsubscribe();
+
+                respond(client, {
+                    id,
+                    jsonrpc,
+                    result: {
+                        type: "stopped",
+                    },
+                });
+            };
+
+            client.data.handleRequest = async (msg: TRPCClientOutgoingMessage) => {
+                const {id, jsonrpc} = msg;
+                /* istanbul ignore next -- @preserve */
+                if (id === null) {
+                    throw new TRPCError({
+                        code: "BAD_REQUEST",
+                        message: "`id` is required",
+                    });
+                }
+                if (msg.method === "subscription.stop") {
```

Comment:
by the way, bun has builtin support for pub/sub. it might be able to be used here. it would be more efficient than a userland implementation, assuming it's feature complete enough to be used

https://bun.sh/docs/api/websockets#pub-sub

Reply:
Don't see how this could help.
Every connection should have its own topic, which doesn't sound for me.